File: python/level1/src/main.py
import sys
import os
import subprocess

class Task:
    def __init__(self, name, dependencies=None, commands=None):
        self.name = name
        self.dependencies = dependencies or []
        self.commands = commands or []

    def __repr__(self):
        return f"Task({self.name}, deps={self.dependencies}, cmds={self.commands})"

class TaskRunner:
    def __init__(self):
        self.tasks = {}
        self.first_target = None

    def parse_file(self, filepath):
        if not os.path.exists(filepath):
            print(f"Error: File '{filepath}' not found.")
            sys.exit(1)

        with open(filepath, 'r') as f:
            current_task = None
            for line in f:
                line = line.rstrip()
                if not line or line.startswith('#'):
                    continue

                if line[0] in ('\t', ' '):
                    # Command (indented line)
                    if current_task:
                        command = line.strip()
                        if command:
                            self.tasks[current_task].commands.append(command)
                    continue

                if ':' in line:
                    # Target: Dependencies
                    parts = line.split(':')
                    target = parts[0].strip()
                    deps = [d.strip() for d in parts[1].split() if d.strip()]
                    
                    self.tasks[target] = Task(target, deps)
                    if self.first_target is None:
                        self.first_target = target
                    current_task = target

    def resolve_dependencies(self, target, visited=None, stack=None):
        if visited is None:
            visited = set()
        if stack is None:
            stack = set()

        if target in stack:
            print(f"Error: Circular dependency detected involving '{target}'")
            sys.exit(1)
        
        if target in visited:
            return []

        if target not in self.tasks:
            # It might be a file on disk without a specific task
            if os.path.exists(target):
                visited.add(target)
                return []
            else:
                print(f"Error: No rule to make target '{target}'")
                sys.exit(1)

        stack.add(target)
        order = []
        for dep in self.tasks[target].dependencies:
            order.extend(self.resolve_dependencies(dep, visited, stack))
        
        stack.remove(target)
        visited.add(target)
        order.append(target)
        return order

    def execute(self, target_order):
        for target_name in target_order:
            task = self.tasks.get(target_name)
            if not task:
                continue
            
            for cmd in task.commands:
                print(f"Executing: {cmd}")
                sys.stdout.flush()
                result = subprocess.run(cmd, shell=True)
                if result.returncode != 0:
                    print(f"Error: Command '{cmd}' failed with exit code {result.returncode}")
                    sys.exit(result.returncode)

def main():
    args = sys.argv[1:]
    taskfile = "Taskfile"
    target = None

    if len(args) >= 1:
        # Check if first arg is a file
        if os.path.exists(args[0]):
            taskfile = args[0]
            if len(args) >= 2:
                target = args[1]
        else:
            target = args[0]

    runner = TaskRunner()
    runner.parse_file(taskfile)
    
    if target is None:
        target = runner.first_target
    
    if target not in runner.tasks and not os.path.exists(target):
        print(f"Error: Target '{target}' not found in '{taskfile}'")
        sys.exit(1)

    order = runner.resolve_dependencies(target)
    runner.execute(order)

if __name__ == "__main__":
    main()

File: golang/level1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type Task struct {
	Name         string
	Dependencies []string
	Commands     []string
}

type TaskRunner struct {
	Tasks       map[string]*Task
	FirstTarget string
}

func NewTaskRunner() *TaskRunner {
	return &TaskRunner{
		Tasks:       make(map[string]*Task),
		FirstTarget: "",
	}
}

func (tr *TaskRunner) parseFile(taskfile string) error {
	absPath, err := filepath.Abs(taskfile)
	if err != nil {
		return err
	}

	if _, err := os.Stat(absPath); err != nil {
		return fmt.Errorf("Error: File '%s' not found.", absPath)
	}

	file, err := os.Open(absPath)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	currentTask := ""
	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r\n")
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		if len(line) > 0 && (line[0] == '\t' || line[0] == ' ') {
			// Command (indented line)
			if currentTask != "" {
				cmd := strings.TrimSpace(line)
				if cmd != "" {
					tr.Tasks[currentTask].Commands = append(tr.Tasks[currentTask].Commands, cmd)
				}
			}
			continue
		}

		if strings.Contains(line, ":") {
			// Target: Dependencies
			parts := strings.SplitN(line, ":", 2)
			target := strings.TrimSpace(parts[0])
			deps := strings.Fields(parts[1])

			tr.Tasks[target] = &Task{
				Name:         target,
				Dependencies: deps,
			}
			if tr.FirstTarget == "" {
				tr.FirstTarget = target
			}
			currentTask = target
		}
	}

	if err := scanner.Err(); err != nil {
		return err
	}

	return nil
}

func (tr *TaskRunner) resolveDependencies(target string, visited map[string]bool, stack map[string]bool) ([]string, error) {
	if target == "" {
		return []string{}, nil
	}

	if stack[target] {
		return []string{}, fmt.Errorf("Error: Circular dependency detected involving '%s'", target)
	}

	visited[target] = true
	stack[target] = true

	task := tr.Tasks[target]
	if task == nil {
		return []string{}, fmt.Errorf("Error: Unknown target '%s'", target)
	}

	var order []string
	for _, dep := range task.Dependencies {
		deps, err := tr.resolveDependencies(dep, visited, stack)
		if err != nil {
			return []string{}, err
		}
		order = append(order, deps...)
	}

	// Add current task after dependencies
	order = append(order, target)

	// Remove cycle tracking for current task
	delete(stack, target)

	return order, nil
}

func (tr *TaskRunner) execute(targetOrder []string) error {
	for _, targetName := range targetOrder {
		task := tr.Tasks[targetName]
		if task == nil {
			continue
		}

		for _, cmd := range task.Commands {
			fmt.Printf("Executing: %s\n", cmd)
			out, err := exec.Command("sh", "-c", cmd).CombinedOutput()
			if err != nil {
				fmt.Printf("Error: Command '%s' failed\n", cmd)
				os.Exit(1)
			}
			if string(out) != "" {
				fmt.Print(out)
			}
		}
	}

	return nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: program <taskfile> [target]")
		os.Exit(1)
	}

	taskfile := os.Args[1]
	
	// Change to taskfile directory before parsing
	taskfileDir := filepath.Dir(taskfile)
	os.Chdir(taskfileDir)
	
	tr := NewTaskRunner()
	if err := tr.parseFile(taskfile); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	target := ""
	if len(os.Args) > 2 {
		target = os.Args[2]
	}

	// If no target specified, use first target
	if target == "" {
		target = tr.FirstTarget
	}

	order, err := tr.resolveDependencies(target, make(map[string]bool), make(map[string]bool))
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	if err := tr.execute(order); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
